draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+referred to as an object. An actor model is a system where objects pass messages
+between one another. CapTP enables objects to have remote references to other
+objects on other CapTP sessions, often on different machines across the network.
+An object with a reference to another object can use it by sending a message or
+sharing the reference in a message to other objects it has references to.
+
+With the right tooling, many programming languages implementing CapTP can
+achieve the same semantics of asynchronous programming for invoking both local
+and remote objects, allowing programmers to focus on code flow over networking
+infrastructure.
+
+# Additional Documents
+
+This document does not stand alone, it relies upon multiple other documents
+which together build up OCapN (Object Capability Network) specifications.
+
+This specification uses the following other specifications:
+
+-   [Syrup](): A serialization format which all correspondences are serialized
+    to and from when crossing a CapTP boundary.
+-   [OCapN Netlayers](): Specification to open a secure communication channel
+    between two sessions, often on different networks.
+-   [OCapN Locators](): Specification covers representation of object references
+    for both in band and out of band usage.
+
+# CapTP Overview
+
+CapTP supports many features, as described in the [Introduction](#introduction)
+section. It can, however, be broken down into the life cycle of a CapTP session.
+A session is between two entities which are communicating via CapTP through a
+OCapN Netlayer's channel. A CapTP life cycle consists of the following stages:
+
+1. A session is established pairwise between machines.
+2. Communication occurs between objects.
+    -   Initial connectivity to objects is established by querying the bootstrap
+        object for access to known objects (via sturdyrefs or certificates).
+    -   Messages are exchanged between objects which hold references to each
+        other.
+    -   Promises may be created when messages are sent.
+    -   Messages may be pipelined to promises, queueing those messages to
+        eventually be delivered to their resolution.
+    -   Handoffs are initiated when sending a message with a reference to an
+        object on a third CapTP session.
+    -   Both parties cooperate to free object references which are recognized to
+        no longer be needed to be imported from one side.
+5.  A session ends.
+
+## Establishing a connection
+
+A new connection is established either by:
+
+- Using an out-of-band mechanism to connect to a remote session
+- Peforming a handoff to connect to a third session
+
+In either situation, a secure channel needs to be created to the remote session.
+This is out of scope for this specification but it's covered by the OCapN
+Netlayers specification. Once a secure channel is established, we MUST to
+perform the following in this order:
+
+1.  Create a per-session cryptography key pair (see
```

Comment:
I wonder why a session key pair is in this layer.

The Agoric CapTP API doesn't use a session key pair.

cc @erights @kriskowal @michaelfig

Reply:
The 3PH presentation from a couple meetings ago goes over how they're used, but I do think it might be good to add some abstraction boundaries; probably most of the crypto should be netlayer specific.

Reply:
I promoted this thread to an issue: #53.